pkg/controllers/job/job_controller.go

```diff
@@ -327,6 +359,37 @@ func (cc *jobcontroller) processNextReq(count uint32) bool {
 
 	klog.V(3).Infof("Try to handle request <%v>", req)
 
+	// Handle cancellation logic for delayed actions:
+	// 1. When request event is not Pod Pending:
+	//   - Cancel corresponding Pod Pending delayed action
+	//   - When Pod enters Running state, cancel corresponding Pod Failed and Pod Evicted delayed actions
+	if req.Event != busv1alpha1.PodPendingEvent {
+		cc.delayActionMapLock.Lock()
+		if taskMap, exists := cc.delayActionMap[key]; exists {
+			for podName, delayAct := range taskMap {
+				shouldCancel := false
+
+				if podName == req.PodName {
+					if delayAct.event == busv1alpha1.PodPendingEvent {
+						shouldCancel = true
+					}
+
+					if req.Event == busv1alpha1.PodRunningEvent &&
+						(delayAct.event == busv1alpha1.PodFailedEvent || delayAct.event == busv1alpha1.PodEvictedEvent) {
+						shouldCancel = true
+					}
+				}
+
+				if shouldCancel {
+					klog.V(3).Infof("Cancel delayed action <%v> for pod <%s> of Job <%s>", delayAct.action, podName, delayAct.jobKey)
+					delayAct.cancel()
+					delete(taskMap, podName)
+				}
+			}
+		}
+		cc.delayActionMapLock.Unlock()
```

Comment:
refactor into a function and unlock by `defer`

Reply:
updated